File: bitcoin_stockmem/pipeline/step5_retrieve.py
"""Step 5: Historical Sequence Retrieval.

Two-stage retrieval:
  1. Coarse: Jaccard-based SeqSim over binary event vectors.
  2. Fine: LLM_retrieve filters candidates to genuine references.
"""

from __future__ import annotations

import json
import logging
from datetime import datetime, timedelta

from config import WINDOW_SIZE, TOP_K_RETRIEVE
from llm.gemini_client import GeminiClient
from llm.prompts import RETRIEVE_PROMPT
from llm.response_parser import parse_retrieve_result
from memory.event_memory import (
    build_event_series, format_series_for_prompt, compute_and_store_daily_vectors,
)
from memory.reflection_memory import (
    get_reflection_by_window, format_reflections_for_prompt,
)
from memory.similarity import find_top_k_sequences
from storage.database import get_cursor

logger = logging.getLogger(__name__)


def _get_all_available_dates(before_date: str, asset: str) -> list[str]:
    """Get all dates that have merged events, before the given date."""
    with get_cursor() as cur:
        cur.execute(
            """SELECT DISTINCT date FROM merged_events
               WHERE date < ? AND (asset = ? OR asset = 'ALL')
               ORDER BY date""",
            (before_date, asset),
        )
        return [row["date"] for row in cur.fetchall()]


def retrieve_references(
    client: GeminiClient,
    date: str,
    asset: str,
) -> tuple[list[dict], list[int]]:
    """Retrieve historically similar event sequences and their reflections.

    Args:
        client: Gemini client.
        date: Current date (end of query window).
        asset: "BTC" or "ETH".

    Returns:
        (reflections_list, reflection_ids) for the selected references.
    """
    # Ensure daily vectors are computed
    end_dt = datetime.strptime(date, "%Y-%m-%d")
    current_dates = []
    for i in range(WINDOW_SIZE, -1, -1):
        d = (end_dt - timedelta(days=i)).strftime("%Y-%m-%d")
        compute_and_store_daily_vectors(d, asset)
        current_dates.append(d)

    # Also ensure historical vectors exist
    all_hist_dates = _get_all_available_dates(current_dates[0], asset)
    for d in all_hist_dates:
        compute_and_store_daily_vectors(d, asset)

    if len(all_hist_dates) < WINDOW_SIZE:
        logger.info(f"Not enough history for retrieval on {date}/{asset}")
        return [], []

    # Stage 1: Coarse — Jaccard-based Top-K
    top_sequences = find_top_k_sequences(
        current_dates, all_hist_dates, asset, k=TOP_K_RETRIEVE
    )

    if not top_sequences:
        logger.info(f"No similar sequences found for {date}/{asset}")
        return [], []

    # Build candidate info with their reflections
    current_dates_list, current_events = build_event_series(date, WINDOW_SIZE, asset)
    current_text = format_series_for_prompt(current_dates_list, current_events)

    candidates_text = []
    candidate_reflections = []
    for i, (hist_dates, sim_score) in enumerate(top_sequences):
        ref = get_reflection_by_window(hist_dates[-1], asset)
        if not ref:
            # Try cross-asset matching
            other_asset = "ETH" if asset == "BTC" else "BTC"
            ref = get_reflection_by_window(hist_dates[-1], other_asset)

        candidate_reflections.append(ref)

        hist_dates_list, hist_events = build_event_series(
            hist_dates[-1], WINDOW_SIZE, asset
        )
        hist_text = format_series_for_prompt(hist_dates_list, hist_events)

        entry = f"\n--- Candidate {i} (similarity: {sim_score:.3f}) ---\n"
        entry += f"Period: {hist_dates[0]} to {hist_dates[-1]}\n"
        entry += hist_text
        if ref:
            entry += f"\nOutcome: price went {ref['price_direction']}"
            entry += f"\nAnalysis: {ref['reason'][:300]}"
        candidates_text.append(entry)

    # Stage 2: Fine — LLM filters
    prompt = RETRIEVE_PROMPT.format(
        window=WINDOW_SIZE,
        asset=asset,
        current_series_text=current_text,
        candidates_with_reflections="\n".join(candidates_text),
    )

    try:
        result = client.generate_json(prompt)
        parsed = parse_retrieve_result(result)
        selected_indices = parsed.selected_indices
    except Exception as e:
        logger.warning(f"Retrieval LLM failed for {date}/{asset}: {e}")
        # Fallback: use top-3 by similarity
        selected_indices = list(range(min(3, len(top_sequences))))

    # Collect selected reflections
    selected_reflections = []
    selected_ids = []
    for idx in selected_indices:
        if 0 <= idx < len(candidate_reflections):
            ref = candidate_reflections[idx]
            if ref:
                selected_reflections.append(ref)
                selected_ids.append(ref["id"])

    logger.info(
        f"{date}/{asset}: retrieved {len(selected_reflections)} references "
        f"from {len(top_sequences)} candidates"
    )
    return selected_reflections, selected_ids

File: bitcoin-stockmem-ts/src/pipeline/step5-retrieve.ts
/**
 * Step 5: Historical Sequence Retrieval.
 * Coarse (Jaccard) + Fine (LLM) two-stage retrieval.
 */

import { GeminiClient } from "../llm/gemini-client";
import { RETRIEVE_PROMPT, fillTemplate } from "../llm/prompts";
import { parseRetrieveResult } from "../llm/response-parser";
import { WINDOW_SIZE, TOP_K_RETRIEVE } from "../config";
import { getDb } from "../storage/database";
import {
  buildEventSeries, formatSeriesForPrompt, computeAndStoreDailyVectors,
} from "../memory/event-memory";
import { getReflectionByWindow, formatReflectionsForPrompt } from "../memory/reflection-memory";
import { findTopKSequences } from "../memory/similarity";

function getAllAvailableDates(beforeDate: string, asset: string): string[] {
  const d = getDb();
  const rows = d.prepare(
    `SELECT DISTINCT date FROM merged_events
     WHERE date < ? AND (asset = ? OR asset = 'ALL') ORDER BY date`
  ).all(beforeDate, asset) as any[];
  return rows.map((r: any) => r.date);
}

export async function retrieveReferences(
  client: GeminiClient,
  date: string,
  asset: string
): Promise<{ reflections: Record<string, any>[]; refIds: number[] }> {
  // Ensure daily vectors exist
  const endDt = new Date(date);
  const currentDates: string[] = [];
  for (let i = WINDOW_SIZE; i >= 0; i--) {
    const d = new Date(endDt);
    d.setDate(d.getDate() - i);
    const ds = d.toISOString().slice(0, 10);
    computeAndStoreDailyVectors(ds, asset);
    currentDates.push(ds);
  }

  const allHistDates = getAllAvailableDates(currentDates[0], asset);
  for (const d of allHistDates) computeAndStoreDailyVectors(d, asset);

  if (allHistDates.length < WINDOW_SIZE) {
    console.log(`Not enough history for retrieval on ${date}/${asset}`);
    return { reflections: [], refIds: [] };
  }

  // Stage 1: Coarse
  const topSeqs = findTopKSequences(currentDates, allHistDates, asset, TOP_K_RETRIEVE);
  if (!topSeqs.length) return { reflections: [], refIds: [] };

  // Build candidates text
  const { dates: curDates, eventsPerDay: curEvents } = buildEventSeries(date, WINDOW_SIZE, asset);
  const currentText = formatSeriesForPrompt(curDates, curEvents);

  const candidateTexts: string[] = [];
  const candidateReflections: (Record<string, any> | null)[] = [];

  for (const [i, seq] of topSeqs.entries()) {
    let ref = getReflectionByWindow(seq.dates[seq.dates.length - 1], asset);
    if (!ref) {
      const other = asset === "BTC" ? "ETH" : "BTC";
      ref = getReflectionByWindow(seq.dates[seq.dates.length - 1], other);
    }
    candidateReflections.push(ref);

    const { dates: hDates, eventsPerDay: hEvents } = buildEventSeries(
      seq.dates[seq.dates.length - 1], WINDOW_SIZE, asset
    );
    const hText = formatSeriesForPrompt(hDates, hEvents);

    let entry = `\n--- Candidate ${i} (similarity: ${seq.score.toFixed(3)}) ---\n`;
    entry += `Period: ${seq.dates[0]} to ${seq.dates[seq.dates.length - 1]}\n`;
    entry += hText;
    if (ref) {
      entry += `\nOutcome: price went ${ref.price_direction}`;
      entry += `\nAnalysis: ${String(ref.reason).slice(0, 300)}`;
    }
    candidateTexts.push(entry);
  }

  // Stage 2: Fine (LLM)
  const prompt = fillTemplate(RETRIEVE_PROMPT, {
    window: String(WINDOW_SIZE),
    asset,
    current_series_text: currentText,
    candidates_with_reflections: candidateTexts.join("\n"),
  });

  let selectedIndices: number[];
  try {
    const result = await client.generateJson(prompt);
    selectedIndices = parseRetrieveResult(result).selected_indices;
  } catch {
    selectedIndices = Array.from({ length: Math.min(3, topSeqs.length) }, (_, i) => i);
  }

  const reflections: Record<string, any>[] = [];
  const refIds: number[] = [];
  for (const idx of selectedIndices) {
    const ref = candidateReflections[idx];
    if (ref) {
      reflections.push(ref);
      refIds.push(ref.id);
    }
  }

  console.log(`${date}/${asset}: retrieved ${reflections.length} refs from ${topSeqs.length} candidates`);
  return { reflections, refIds };
}
